src/controllers/google.auth.controller.ts

```diff
@@ -0,0 +1,94 @@
+import passport from 'passport';
+import bcrypt from 'bcrypt';
+import { User } from '../models/index';
+import dotenv from 'dotenv';
+import { Request } from 'express';
+
+dotenv.config();
+const GoogleStrategy = require('passport-google-oauth2').Strategy; // eslint-disable-line
+
+interface GoogleProfile {
+  id: string;
+  displayName: string;
+  name: { familyName: string; givenName: string };
+  email: { value: string; verified: boolean }[];
+  photos: { value: string }[];
+}
+
+const GOOGLE_CLIENT_ID = process.env.GOOGLE_CLIENT_ID;
+const GOOGLE_CLIENT_SECRET = process.env.GOOGLE_CLIENT_SECRET;
+const GOOGLE_CALLBACK_URL = process.env.GOOGLE_CALLBACK_URL;
+passport.use(
+  new GoogleStrategy(
+    {
+      clientID: GOOGLE_CLIENT_ID,
+      clientSecret: GOOGLE_CLIENT_SECRET,
+      callbackURL: GOOGLE_CALLBACK_URL,
+      passReqToCallback: true,
+    },
+    async function (
+      request: Request,
+      accessToken: unknown,
+      refreshToken: unknown,
+      profile: GoogleProfile,
+      done: (arg0: unknown, arg1: unknown) => unknown,
+    ) {
+      try {
+        let user = await User.findOne({ where: { email: profile.email } });
+        if (!request.res) {
+          throw new Error('Response not found');
+        }
+        if (user) {
+          const token = accessToken;
+          request.res
+            .header('Authorization', `Bearer ${token}`)
+            .cookie('access_token', token, {
+              httpOnly: true,
+              maxAge: 60 * 600 * 1000,
+            });
+          done(null, { user, token });
+        } else {
+          const hashedPassword = await bcrypt.hash(profile.id, 10);
+          const hashedGoogleId = await bcrypt.hash(profile.id, 10);
+          user = await User.create({
+            username: profile.displayName,
+            email: profile.email,
+            password: hashedPassword,
+            firstname: profile.name.givenName,
+            lastname: profile.name.familyName,
+            role: 'USER',
+            verified: true,
+            image_url: profile.photos[0].value,
+            google_id: hashedGoogleId,
+            google_token: accessToken,
+          });
+          const token = accessToken;
+          request.res
+            .header('Authorization', `Bearer ${token}`)
+            .cookie('access_token', token, {
+              httpOnly: true,
+              maxAge: 60 * 600 * 1000,
+            });
+
+          done(null, { user, token });
+        }
+      } catch (error: unknown) {
+        return done(error, null);
+      }
+    },
+  ),
+);
+
+passport.serializeUser(function (
+  user,
+  done: (arg0: null, arg1: unknown) => void,
+) {
+  done(null, user);
+});
+
+passport.deserializeUser(function (
+  user,
+  done: (arg0: null, arg1: unknown) => void,
+) {
+  done(null, user);
+});
```

Comment:
This is not a controller, this should be a middleware such as passport.middleware.ts for example... then after that you can use it in a googleAuth controller where you can implement a function for googleLogin

Reply:
> This is not a controller, this should be a middleware such as passport.middleware.ts for example... then after that you can use it in a googleAuth controller where you can implement a function for googleLogin

Thanks, but I have put this into config, since it will be used to only configure this passport strategy, and it has no direct route connected to it, but instead loading this strategy in the main server.ts file so that whenever our server runs, it loads this also. It is not being used in this server.ts file. It is only importing it there. That is why I preferred making it a config file.

Reply:
Yes but we still need a controller to use while logging in?? He's what you need to do. You need to make a controller specifically for logging in that will first check whether the user exists by getting the credential from the body and passing it to the `client.verifyIdToken` from ouath2. If the verification doesn't work then give them an error, if it does then extract their data from the payload and use it to find out whether they exist in our db. If they don't exist in our db then they should signup, if they exist then generate a token for them, in this case a cookie and set it in their header like we're doing. Then you shall call this controller in the passport.routes where we shall have all other routes for passport as well. @Aldot-02